ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -227,123 +317,160 @@ prune maxSize st
---
--- = Properties
---
--- * Rewinding to the last applied block is an identity
---
---   > rewind (lastSlot st) st == Just st
---
--- * Rewinding to the same slot should be idempotent:
---
---   > rewind s st == Just st'  ==>  rewind s st' == Just st'
+-- implicitly applied).
 --
--- * Rewinding fails only if the slot is before the anchor
---
---   > rewind s st == Nothing  iff  s < anchor st
---
---   (See documentation of 'anchor' for motivation)
---
--- * If rewinding to a particular slot fails, rewinding to any slot prior to
---   that will also fail
---
---   > rewind s st == Nothing, s' < s  ==>  rewind s' st == Nothing
---
--- * Rewinding to the anchor will leave the state empty
---
---   > size (rewind (anchor st) st) == 0
---
--- * But rewinding to a more recent block will not
---
---   > s > anchor st  ==>  size (rewind s st) > 0
-rewind :: PBftCrypto c
-       => WithOrigin SlotNo -> PBftChainState c -> Maybe (PBftChainState c)
-rewind slot st
-  | slot < anchor st = Nothing
-  | otherwise = Just $ PBftChainState {
-        anchor  = anchor st
-      , signers = keep
-      -- Typically we only drop a few blocks (and keep most), so updating
-      -- the counts based on what we drop will in most cases be faster than
-      -- recomputing it from the blocks we keep.
-      , counts  = repeatedly (decrementKey . pbftSignerGenesisKey)
-                             (toList discard)
-                             (counts st)
-      }
+-- In addition to preserving the invariant, we also have the guarantee that
+-- rolling back to a point (within @k@) and then reapplying the blocks that were
+-- rolled back results in the original state.
+rewind :: forall c. PBftCrypto c
+       => SecurityParam
+       -> WindowSize
+       -> WithOrigin SlotNo
+       -> PBftChainState c -> Maybe (PBftChainState c)
+rewind k n mSlot PBftChainState{..} =
+    case mSlot of
+      At slot ->
+        -- We scan from the right, since block to roll back to likely at end
+        case Seq.spanr (\(PBftSigner slot' _) -> slot' > slot) postAnchor of
+
+          -- We found the slot to roll back to post-anchor. Discard everything
+          -- after that slot.
+          (toDiscard, toKeep@(_ :|> x)) ->
+            if slot == pbftSignerSlotNo x
+              then Just $ go toDiscard toKeep
+              else notPreviouslyApplied
+
+          -- The slot was not found post-anchor. If the slot matches the last
+          -- slot pre-anchor, all is well, discarding everything post-anchor.
+          -- Otherwise, the rollback is too far.
+          (toDiscard, Empty) ->
+            case preAnchor of
+              _ :|> x
+                | slot == pbftSignerSlotNo x -> Just $ go toDiscard Empty
+                | slot <  pbftSignerSlotNo x -> rollbackTooFar
+                | otherwise                  -> notPreviouslyApplied
+              _otherwise ->
+                notPreviouslyApplied
```

Comment:
If we still have an incomplete `ChainState` because we restarted from an old one, a valid call to `rewindChainState` might trigger this case, resulting in an `error` instead of `Nothing`.

When exactly would this happen?
1. The node is running a version of the code before this PR.
2. We roll back our chain `r` (<= `k`) blocks such that our window is no longer filled, it now contains `n - r` signers. Note that `n = k`.
3. The node is shut down.
4. We restart the node but now using a version of the code after this PR. Now the `ChainState` is incomplete, we have `n - r` signers. The `preAnchor` is empty and `postAnchor` contains `n - r` signers (correct?).
5. The ChainSync client now connects to a server that is `x` blocks back with `x >= n - r`. We need to rewind our `ChainState` `x` blocks, including the signers window. We drop `x` signers and thus all signers from the `postAnchor`. We encounter an empty `preAnchor` and call `notPreviouslyApplied`, which throws an `error`, because our `preAnchor` has not yet been sufficiently populated.
6. The node won't shut down, just the connection to the server (right? I believe that's what the error policy now does), which is running behind anyway.

The chances of this happening are slim: (2) is rare, as rollbacks are rare, (5) is also rather unlikely, as it is unlikely that there is a server that many blocks behind. Certainly after *we* restarted, because I would expect *us* to be running behind.

So I think we can leave this as is.

Reply:
I'm happy with this analysis and agree with the conclusion.

Reply:
New insight after talking with @dcoutts:

The `ChainState` is restored from a ledger snapshot. We will use a ledger snapshot that is not newer than the tip of the ImmutableDB to restore from, and will then typically replay `k` or more blocks against that ledger snapshot to bring it up to date with the tip of the chain (include the volatile part). This replay of `k` or more blocks is exactly what we need to fill our incomplete signer window again (we need `k` blocks)! Only when we're near genesis or in case of VolatileDB corruption, meaning that we will replay less than `k` blocks, do we still end up with an incomplete signer window.